pkg/osinfo/scanner_rpm.go

```diff
@@ -0,0 +1,136 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package osinfo
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	// Import sqlite driver for rpm database
+	_ "github.com/glebarez/go-sqlite"
+	rpmdbpkg "github.com/knqyf263/go-rpmdb/pkg"
+	"github.com/sirupsen/logrus"
+)
+
+type rpmScanner struct {
+	ls layerScanner
+}
+
+func newRPMScanner() containerOSScanner {
+	return &rpmScanner{
+		ls: newLayerScanner(),
+	}
+}
+
+func (ct *rpmScanner) PURLType() string {
+	return "rpm"
+}
+
+func (ct *rpmScanner) OSType() OSType {
+	return OSRHEL
+}
+
+// ReadOSPackages reads the rpm database
+func (ct *rpmScanner) ReadOSPackages(layers []string) (layer int, pk *[]PackageDBEntry, err error) {
+	rpmDatabase := ""
+
+	rpmDBFiles := []string{
+		"rpmdb.sqlite", // sqlite
+		"Packages.db",  // ndb
+		"Packages",     // BerkleyDB
+	}
+
+	for i, lp := range layers {
+		tmpDBdir, err := os.MkdirTemp("", "rmpdb")
+		defer os.RemoveAll(tmpDBdir)
+		if err != nil {
+			return 0, pk, fmt.Errorf("creating temporary rpmdb dir: %w", err)
+		}
+		for _, dbname := range rpmDBFiles {
+			tmpDBPath := filepath.Join(tmpDBdir, dbname)
+			rpmdbpath := filepath.Join("var/lib/rpm", dbname)
+			exists, err := ct.ls.FileExistsInTar(lp, rpmdbpath)
+			if err != nil {
+				return 0, pk, fmt.Errorf("extracting rpm database: %w", err)
+			}
+			if exists {
+				err := ct.ls.ExtractFileFromTar(lp, rpmdbpath, tmpDBPath)
+				if err != nil {
+					os.Remove(tmpDBPath)
+					if _, ok := err.(ErrFileNotFoundInTar); ok {
+						continue
+					}
+					return 0, pk, fmt.Errorf("extracting rpm database: %w", err)
+				}
+				logrus.Debugf("Layer %d has a newer version of rpm database", i)
+				rpmDatabase = tmpDBPath
+				layer = i
+				// Skip to the next layer. A single layer shouldn't have multiple
+				// database formats in it. Even if it did, we stop for the most new
+				// format present in the layer
+				goto NEXT_LAYER
+			}
+		}
+	NEXT_LAYER:
+	}
+
+	if rpmDatabase == "" {
+		logrus.Info("rpm database data is empty")
+		return layer, nil, nil
+	}
+
+	pk, err = ct.ParseDB(rpmDatabase)
+	if err != nil {
+		return layer, nil, fmt.Errorf("parsing rpm database: %w", err)
+	}
+	return layer, pk, err
+}
+
+func (ct *rpmScanner) ParseDB(dbPath string) (*[]PackageDBEntry, error) {
+	db, err := rpmdbpkg.Open(dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("opening rpmdb: %w", err)
+	}
+
+	pkgs, err := db.ListPackages()
+	if err != nil {
+		return nil, fmt.Errorf("parsing rpm db: %w", err)
+	}
+	packages := []PackageDBEntry{}
+	for _, p := range pkgs {
+		if _, ok := virtualPackages[p.Name]; ok {
+			continue
+		}
+
+		packages = append(packages, PackageDBEntry{
+			Package:      p.Name,
+			Version:      p.Version,
+			Architecture: p.Arch,
+			Type:         "rpm",
+			// Namespace is set later
+			MaintainerName: p.Vendor,
+			// Most RPM pacakges don't have SPDX-valid license names
```

Comment:
Curious: Is this true for all distros? Are they using an alternative naming scheme that could be translated later?

Reply:
This potentially affects other distros, I haven't looked at APK or DEB for validation. I spent a _lot_ of time digging into this and found that basically the story is not great:
* If you stuff a SPDX-invalid license name into the field, the SPDX doc fails to verify
* Some distros like [Fedora](https://fedoraproject.org/wiki/Changes/SPDX_Licenses_Phase_2) are working to migrate license names to SPDX license strings. This effort is not complete yet for Fedora.  Fedora has a [`license-fedora2spdx`](https://pagure.io/copr/license-validate) package (Python scripts/ config) that tries to do convert license name strings, but fails pretty hard in a lot of cases. Ex: `BSD` was Fedora-valid, but the tool spits out something like `"BSD": "Warning: more options on how to interpret BSD. Possible options: ['BSD-1-Clause', 'BSD-2-Clause-FreeBSD', 'BSD-2-Clause-Views', 'BSD-2-Clause', 'BSD-3-Clause-Modification', 'BSD-3-Clause']` which requires more info. 
* There isn't a documented way to parse the RPM database to produce the equivalent of `rpm -q --licensefiles $PACKAGE` and get the list of license files in the image. 
  * Most distros seem to put license files in `/usr/share/licenses/$PACKAGE` with multiple different filenames (Try running `find /usr/share/licenses -type f -exec basename {} \; | sort |uniq | wc -l` in any RPM-based container distro, you'll get >40 different names). Even if you ran `rpm -qL $PKG`, or just searched in the directory and tried to analyze the license files, you wouldn't know if there's an `AND` vs an `OR` for a given package. (`AND` typically indicating valid SPDX license vs `and` being a legacy license name. Try `rpm -qa --queryformat '%{name} %{license}\n' | grep -E ' (and|or) '` to find those legacy ones. )

We _could_ try calling [`Reader.LicenseFromLabel()`](https://github.com/micahhausler/bom/blob/rpm-scanner/pkg/license/license.go#L153:L153) and only including valid ones, but for now I just mirrored what DEB and APK parsers do.